### PR TITLE
Fix #2263: Fixes App crashes in Lessons Tab and Espresso tests in TopicLessonFragmentTest

### LINKED ISSUE #2263: Lessons tab crashes in landscape mode(mobile) and tablets
**Describe the bug**
Lessons tab crash when switched to landscape.

**To Reproduce**
1. Open Any Topic
2. open on 'Lessons Tab'
3. Change to Landscape mode
4. App Crashes

**Expected behavior**
Should not crash on landscape mode

**Screenshots**
If applicable, add screenshots or videos to help explain your problem.

**Device**
- Pixel 3a XL - emulator
- SDK version - 19

**Additional context**

This crash is introduced in the PR #2104 

error on crash
```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.oppia.android, PID: 28818
    java.lang.IllegalStateException: Trying to bind incompatible data to adapter. Data class type: class org.oppia.android.app.model.ChapterSummary, expected adapter class type: class org.oppia.android.app.topic.lessons.ChapterSummaryViewModel.
        at org.oppia.android.app.recyclerview.BindableAdapter.setDataUnchecked(BindableAdapter.kt:62)
        at org.oppia.android.app.recyclerview.RecyclerViewBindingAdapter.bindToRecyclerViewAdapter(RecyclerViewBindingAdapter.java:68)
        at org.oppia.android.app.recyclerview.RecyclerViewBindingAdapter.bindToRecyclerViewAdapterWithoutLiveData(RecyclerViewBindingAdapter.java:38)
        at org.oppia.android.databinding.TopicLessonsStorySummaryBindingLandImpl.executeBindings(TopicLessonsStorySummaryBindingLandImpl.java:257)
        at androidx.databinding.ViewDataBinding.executeBindingsInternal(ViewDataBinding.java:473)
        at androidx.databinding.ViewDataBinding.executePendingBindings(ViewDataBinding.java:445)
        at androidx.databinding.ViewDataBinding$7.run(ViewDataBinding.java:197)
        at androidx.databinding.ViewDataBinding$8.doFrame(ViewDataBinding.java:291)
        at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1029)
        at android.view.Choreographer.doCallbacks(Choreographer.java:854)
        at android.view.Choreographer.doFrame(Choreographer.java:785)
        at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1016)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:224)
        at android.app.ActivityThread.main(ActivityThread.java:7592)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:539)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950)
```

### PR DESCRIPTION
**Pull request by FareesHussain**
## Explanation

Fixes #2263 

 - Fixes app crash in lessons Tab (landscape)
 - Fixes Espresso lagging test case in TopicLessonFragmentTest (testLessonsPlayFragment_loadRatiosTopic_configurationChange_storyName_isCorrect)
- Fixes app crash in lesson Tab (Tablet portrait and Tablet landscape)
 - Fixes 12 failed test in TopicLessonFragmentTest when run in TopicLessonFragmentTest (run in tablet)

**NOTE**:  above mentioned tests passes in robolectric

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

##ScreenShots
testLessonsPlayFragment_loadRatiosTopic_configurationChange_storyName_isCorrect before changes 
image
Espresso test failing in tablet
image

### CODE REVIEW

**Comment by FareesHussain** [comment]
Workflow needs to be re-run as it works when I run it in my [forked workflow](https://github.com/FareesHussain/oppia-android/runs/1556276174?check_suite_focus=true)

**Comment by anandwana001** [comment]
@FareesHussain can we investigate why this crash is occurring? As I know earlier it is not crashing.

**Comment by FareesHussain** [comment]
may be bindable adapter was implemented in story summary recently 
the person who implemented it might have forgot to make changes in other layouts

**Comment by anandwana001** [comment]
> may be bindable adapter was implemented in story summary recently
> the person who implemented it might have forgot to make changes in other layouts

Could you link the PR which cause the crash? We can fix but need to check why things working correctly before.

**Comment by FareesHussain** [comment]
> > may be bindable adapter was implemented in story summary recently
> > the person who implemented it might have forgot to make changes in other layouts
> 
> Could you link the PR which cause the crash? We can fix but need to check why things working correctly before.

@anandwana001 It was working correctly because robolectric tests are passing and it was tested in mobile portrait 
and this is the pr #2104

**Comment by anandwana001** [comment]
> #2104

Looking into this, Thanks for mentioning. 

Ah, I really messed up the configuration here by passing the faulty data. 
Wondering why the test was passing on Robolectric!!

Could you run the Bazel tests, steps are listed here - https://github.com/oppia/oppia-android/pull/2104#issuecomment-734540089

**Review by anandwana001 — APPROVED** [review]
LGTM, Thanks @FareesHussain

**Review by rt4914 — APPROVED** [review]
Thanks a lot @FareesHussain

**Comment by rt4914** [comment]
This will need @BenHenning approval as he is the code owner.

**Comment by FareesHussain** [comment]
@anandwana001 I tried to run the Bazel test you mentioned 

for the first command
```bazel build //oppia --android_databinding_use_v3_4_args --experimental_android_databinding_v2 --java_header_compilation=false --noincremental_dexing --define=android_standalone_dexing_tool=d8_compat_dx```

I get the output
```
DEBUG: Rule 'io_bazel_stardoc' indicated that a canonical reproducible form can be obtained by modifying arguments commit = "4378e9b6bb2831de7143580594782f538f461180", shallow_since = "157082916
6 -0400" and dropping ["tag"]
DEBUG: Repository io_bazel_stardoc instantiated at:
  E:/android/oppia-android-two/WORKSPACE:45:35: in <toplevel>
  C:/users/syed farees hussain/_bazel_syed farees hussain/zcnrcatv/external/io_bazel_rules_kotlin/kotlin/internal/repositories/download.bzl:115:10: in kt_download_local_dev_dependencies
  C:/users/syed farees hussain/_bazel_syed farees hussain/zcnrcatv/external/bazel_tools/tools/build_defs/repo/utils.bzl:201:18: in maybe
Repository rule git_repository defined at:
  C:/users/syed farees hussain/_bazel_syed farees hussain/zcnrcatv/external/bazel_tools/tools/build_defs/repo/git.bzl:199:33: in <toplevel>
DEBUG: Rule 'robolectric' indicated that a canonical reproducible form can be obtained by modifying arguments sha256 = "a7eac9418cc0b6d2527699986d5937efcebae927c3917c8bd3cb7663ee65ab63"
DEBUG: Repository robolectric instantiated at:
  E:/android/oppia-android-two/WORKSPACE:91:13: in <toplevel>
Repository rule http_archive defined at:
  C:/users/syed farees hussain/_bazel_syed farees hussain/zcnrcatv/external/bazel_tools/tools/build_defs/repo/http.bzl:336:31: in <toplevel>
DEBUG: Rule 'tools_android' indicated that a canonical reproducible form can be obtained by modifying arguments shallow_since = "1594238320 -0400"
DEBUG: Repository tools_android instantiated at:
  E:/android/oppia-android-two/WORKSPACE:102:15: in <toplevel>
Repository rule git_repository defined at:
  C:/users/syed farees hussain/_bazel_syed farees hussain/zcnrcatv/external/bazel_tools/tools/build_defs/repo/git.bzl:199:33: in <toplevel>
ERROR: Skipping '//oppia': no such package 'oppia': BUILD file not found in any of the following directories. Add a BUILD file to a directory to mark it as a package.
 - E:/android/oppia-android-two/oppia
WARNING: Target pattern parsing failed.
ERROR: no such package 'oppia': BUILD file not found in any of the following directories. Add a BUILD file to a directory to mark it as a package.
 - E:/android/oppia-android-two/oppia
INFO: Elapsed time: 212.216s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded)
```

Could you please check

**Comment by FareesHussain** [comment]
For the second command
```
bazel test //app:TopicLessonsFragmentTest --android_databinding_use_v3_4_args --experimental_android_databinding_v2 --java_header_compilation=false --noincremental_dexing --define=android_standalone_dexing_tool=d8_compat_dx
```
The output is

```
DEBUG: Rule 'io_bazel_stardoc' indicated that a canonical reproducible form can be obtained by modifying arguments commit = "4378e9b6bb2831de7143580594782f538f461180", shallow_since = "1570829166 -0400" and dropping ["tag"]
DEBUG: Repository io_bazel_stardoc instantiated at:
  E:/android/oppia-android-two/WORKSPACE:45:35: in <toplevel>
  C:/users/syed farees hussain/_bazel_syed farees hussain/zcnrcatv/external/io_bazel_rules_kotlin/kotlin/internal/repositories/download.bzl:115:10: in kt_download_local_dev_dependencies
  C:/users/syed farees hussain/_bazel_syed farees hussain/zcnrcatv/external/bazel_tools/tools/build_defs/repo/utils.bzl:201:18: in maybe
Repository rule git_repository defined at:
  C:/users/syed farees hussain/_bazel_syed farees hussain/zcnrcatv/external/bazel_tools/tools/build_defs/repo/git.bzl:199:33: in <toplevel>
DEBUG: Rule 'robolectric' indicated that a canonical reproducible form can be obtained by modifying arguments sha256 = "a7eac9418cc0b6d2527699986d5937efcebae927c3917c8bd3cb7663ee65ab63"
DEBUG: Repository robolectric instantiated at:
  E:/android/oppia-android-two/WORKSPACE:91:13: in <toplevel>
Repository rule http_archive defined at:
  C:/users/syed farees hussain/_bazel_syed farees hussain/zcnrcatv/external/bazel_tools/tools/build_defs/repo/http.bzl:336:31: in <toplevel>
DEBUG: Rule 'tools_android' indicated that a canonical reproducible form can be obtained by modifying arguments shallow_since = "1594238320 -0400"
DEBUG: Repository tools_android instantiated at:
  E:/android/oppia-android-two/WORKSPACE:102:15: in <toplevel>
Repository rule git_repository defined at:
  C:/users/syed farees hussain/_bazel_syed farees hussain/zcnrcatv/external/bazel_tools/tools/build_defs/repo/git.bzl:199:33: in <toplevel>
INFO: Repository maven instantiated at:
  E:/android/oppia-android-two/WORKSPACE:125:14: in <toplevel>
  C:/users/syed farees hussain/_bazel_syed farees hussain/zcnrcatv/external/rules_jvm_external/defs.bzl:89:19: in maven_install
Repository rule coursier_fetch defined at:
  C:/users/syed farees hussain/_bazel_syed farees hussain/zcnrcatv/external/rules_jvm_external/coursier.bzl:620:33: in <toplevel>
INFO: Repository 'maven' used the following cache hits instead of downloading the corresponding file.
 * Hash '8f35f92fb8e021f96b3aa8145c66c3b2e29295baabb28ff50569e613438afcbd' for https://github.com/coursier/coursier/releases/download/v2.0.0-RC3-4/coursier.jar
If the definition of 'maven' was updated, verify that the hashes were also updated.
ERROR: An error occurred during the fetch of repository 'maven':
   Traceback (most recent call last):
        File "C:/users/syed farees hussain/_bazel_syed farees hussain/zcnrcatv/external/rules_jvm_external/coursier.bzl", line 364, column 13, in _coursier_fetch_impl
                fail("Error while fetching artifact with coursier: " + exec_result.stderr)
Error in fail: Error while fetching artifact with coursier: Timed out
ERROR: E:/android/oppia-android-two/app/BUILD.bazel:1092:9: //app:TopicLessonsFragmentTest depends on @maven//:androidx_annotation_annotation in repository @maven which failed to fetch. no such package '@maven//': Error while fetching artifact w
ith coursier: Timed out
ERROR: E:/android/oppia-android-two/app/BUILD.bazel:1092:9: //app:TopicLessonsFragmentTest depends on @maven//:androidx_databinding_databinding_adapters in repository @maven which failed to fetch. no such package '@maven//': Error while fetching
 artifact with coursier: Timed out
ERROR: E:/android/oppia-android-two/app/BUILD.bazel:1092:9: //app:TopicLessonsFragmentTest depends on @maven//:androidx_databinding_databinding_common in repository @maven which failed to fetch. no such package '@maven//': Error while fetching a
rtifact with coursier: Timed out
ERROR: E:/android/oppia-android-two/app/BUILD.bazel:1092:9: //app:TopicLessonsFragmentTest depends on @maven//:androidx_databinding_databinding_compiler in repository @maven which failed to fetch. no such package '@maven//': Error while fetching
 artifact with coursier: Timed out
ERROR: E:/android/oppia-android-two/app/BUILD.bazel:1092:9: //app:TopicLessonsFragmentTest depends on @maven//:androidx_databinding_databinding_runtime in repository @maven which failed to fetch. no such package '@maven//': Error while fetching
artifact with coursier: Timed out
ERROR: E:/android/oppia-android-two/app/BUILD.bazel:1092:9: //app:TopicLessonsFragmentTest depends on @maven//:androidx_navigation_navigation_fragment in repository @maven which failed to fetch. no such package '@maven//': Error while fetching a
rtifact with coursier: Timed out
ERROR: E:/android/oppia-android-two/app/BUILD.bazel:1092:9: //app:TopicLessonsFragmentTest depends on @maven//:androidx_navigation_navigation_ui in repository @maven which failed to fetch. no such package '@maven//': Error while fetching artifac
t with coursier: Timed out
ERROR: E:/android/oppia-android-two/app/BUILD.bazel:1092:9: //app:TopicLessonsFragmentTest depends on @maven//:androidx_test_espresso_espresso_contrib in repository @maven which failed to fetch. no such package '@maven//': Error while fetching a
rtifact with coursier: Timed out
ERROR: E:/android/oppia-android-two/app/BUILD.bazel:1092:9: //app:TopicLessonsFragmentTest depends on @maven//:androidx_test_espresso_espresso_core in repository @maven which failed to fetch. no such package '@maven//': Error while fetching arti
fact with coursier: Timed out
ERROR: E:/android/oppia-android-two/app/BUILD.bazel:1092:9: //app:TopicLessonsFragmentTest depends on @maven//:androidx_test_espresso_espresso_intents in repository @maven which failed to fetch. no such package '@maven//': Error while fetching a
rtifact with coursier: Timed out
ERROR: E:/android/oppia-android-two/app/BUILD.bazel:1092:9: //app:TopicLessonsFragmentTest depends on @maven//:androidx_test_ext_junit in repository @maven which failed to fetch. no such package '@maven//': Error while fetching artifact with cou
rsier: Timed out
ERROR: E:/android/oppia-android-two/app/BUILD.bazel:1092:9: //app:TopicLessonsFragmentTest depends on @maven//:androidx_test_runner in repository @maven which failed to fetch. no such package '@maven//': Error while fetching artifact with coursi
er: Timed out
ERROR: E:/android/oppia-android-two/app/BUILD.bazel:1092:9: //app:TopicLessonsFragmentTest depends on @maven//:com_github_bumptech_glide_mocks in repository @maven which failed to fetch. no such package '@maven//': Error while fetching artifact
with coursier: Timed out
ERROR: E:/android/oppia-android-two/app/BUILD.bazel:1092:9: //app:TopicLessonsFragmentTest depends on @maven//:com_google_truth_truth in repository @maven which failed to fetch. no such package '@maven//': Error while fetching artifact with cour
sier: Timed out
ERROR: E:/android/oppia-android-two/app/BUILD.bazel:1092:9: //app:TopicLessonsFragmentTest depends on @maven//:org_jetbrains_kotlin_kotlin_test_junit in repository @maven which failed to fetch. no such package '@maven//': Error while fetching ar
tifact with coursier: Timed out
ERROR: E:/android/oppia-android-two/app/BUILD.bazel:1092:9: //app:TopicLessonsFragmentTest depends on @maven//:org_jetbrains_kotlin_kotlin_reflect in repository @maven which failed to fetch. no such package '@maven//': Error while fetching artif
act with coursier: Timed out
ERROR: E:/android/oppia-android-two/app/BUILD.bazel:1092:9: //app:TopicLessonsFragmentTest depends on @maven//:org_jetbrains_kotlinx_kotlinx_coroutines_test in repository @maven which failed to fetch. no such package '@maven//': Error while fetc
hing artifact with coursier: Timed out
ERROR: E:/android/oppia-android-two/app/BUILD.bazel:1092:9: //app:TopicLessonsFragmentTest depends on @maven//:org_mockito_mockito_core in repository @maven which failed to fetch. no such package '@maven//': Error while fetching artifact with co
ursier: Timed out
ERROR: E:/android/oppia-android-two/app/BUILD.bazel:1092:9: //app:TopicLessonsFragmentTest depends on @maven//:org_robolectric_annotations in repository @maven which failed to fetch. no such package '@maven//': Error while fetching artifact with
 coursier: Timed out
ERROR: Analysis of target '//app:TopicLessonsFragmentTest' failed; build aborted: Analysis failed
INFO: Elapsed time: 602.467s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded, 0 targets configured)
FAILED: Build did NOT complete successfully (0 packages loaded, 0 targets configured)

```

**Comment by anandwana001** [comment]
@FareesHussain It involves cloning of oppia's bazel version. I will do it for now. 
Please add the screenshot of espresso and robolectric result for future reference.

**Comment by FareesHussain** [comment]
Espresso 
image
Robolectric
image

**Comment by BenHenning** [comment]
I'm fine with approving this, but can we please update #2263 with the root cause (e.g. which PR introduced the original crash)? In general, cases like this should lead to a reversion, not a fix (see https://github.com/oppia/oppia-android/wiki/Revert-&-regression-policy).

**Comment by BenHenning** [comment]
Also, before approving: @FareesHussain please add tests in the fragment holding these views that hit each of these scenarios (e.g. landscape, tablet portrait, tablet landscape). Please add a screenshot showing these tests fail without your fix in place (to ensure they prevent this from happening again).

**Review by BenHenning — CHANGES_REQUESTED** [review]
Requesting changes until there's a test to guard against this regressing in the future.

**Comment by FareesHussain** [comment]
> Requesting changes until there's a test to guard against this regressing in the future.

@BenHenning I've already added them in this pr description please check

**Comment by FareesHussain** [comment]
> Please add a screenshot showing these tests fail without your fix in place (to ensure they prevent this from happening again).

I've already added screenshots in this pr description please check

**Comment by FareesHussain** [comment]
@BenHenning  please check
## Before changes

For mobile landscape the following tests does'nt run to the completions in espresso (however they pass in robolectric)

*testLessonsPlayFragment_loadRatiosTopic_clickExpandListIconIndex1_configurationChange_chapterListIsVisible*
image

*testLessonsPlayFragment_loadRatiosTopic_configurationChange_storyName_isCorrect*
image


*Running all tests of TopicLessonsFragmentTest 12 out of 13 Tests fail*
image

## After Changes
*Mobile espresso*
image

*Tablet Espresso* (NEXUS 7 API 27)
image

**Comment by BenHenning** [comment]
Thanks @FareesHussain. Sorry to be a bit clearer, what we actually need here is a test that results in the same crash that you're fixing. Are those Espresso tests actually resulting in the same bindable adapter issue that #2263 is describing? It's also odd that Robolectric isn't failing--it should be. @FareesHussain can you make sure that we have a test that specifically fails on Robolectric without the suggest fix in place? @anandwana001 might this be because Robolectric doesn't do rotation correctly? If so, @FareesHussain you may be able to write an explicit test for this that starts the device in landscape mode (e.g. using ``@Config`` per http://robolectric.org/javadoc/3.3/org/robolectric/annotation/Config.html#qualifiers--).

**Comment by FareesHussain** [comment]
> Thanks @FareesHussain. Sorry to be a bit clearer, what we actually need here is a test that results in the same crash that you're fixing. Are those Espresso tests actually resulting in the same bindable adapter issue that #2263 is describing? It's also odd that Robolectric isn't failing--it should be. @FareesHussain can you make sure that we have a test that specifically fails on Robolectric without the suggest fix in place? @anandwana001 might this be because Robolectric doesn't do rotation correctly? If so, @FareesHussain you may be able to write an explicit test for this that starts the device in landscape mode (e.g. using `@Config` per http://robolectric.org/javadoc/3.3/org/robolectric/annotation/Config.html#qualifiers--).

@BenHenning  There is no crash when I run this on Espresso(it crashes when the Item in the recyclerview is expanded) The test stop as below. I will try adding tests corresponding test that fail in robolectric.

Screenshot_1608264360

**Comment by anandwana001** [comment]
> Thanks @FareesHussain. Sorry to be a bit clearer, what we actually need here is a test that results in the same crash that you're fixing. Are those Espresso tests actually resulting in the same bindable adapter issue that #2263 is describing? It's also odd that Robolectric isn't failing--it should be. @FareesHussain can you make sure that we have a test that specifically fails on Robolectric without the suggest fix in place? @anandwana001 might this be because Robolectric doesn't do rotation correctly? If so, @FareesHussain you may be able to write an explicit test for this that starts the device in landscape mode (e.g. using `@Config` per http://robolectric.org/javadoc/3.3/org/robolectric/annotation/Config.html#qualifiers--).

Sounds like that. Robolectric is not performing rotation correctly.

**Comment by FareesHussain** [comment]
@BenHenning please check

*Before changes robolectric*
image

*After changes robolectric*
image

**Comment by FareesHussain** [comment]
Workflow needs to be re-run as the checks pass on my forked workflow

https://github.com/FareesHussain/oppia-android/actions/runs/433429817

File: app/src/sharedTest/java/org/oppia/android/app/topic/lessons/TopicLessonsFragmentTest.kt [hunk 1]
```diff
@@ -308,6 +311,7 @@ class TopicLessonsFragmentTest {
   fun testLessonsPlayFragment_loadRatiosTopic_clickExpandListIconIndex1_configurationChange_chapterListIsVisible() { // ktlint-disable max-line-length
     launch<TopicActivity>(createTopicActivityIntent(internalProfileId, RATIOS_TOPIC_ID)).use {
       clickLessonTab()
+      onView(isRoot()).perform(orientationLandscape())
```
Comment (FareesHussain) [inline app/src/sharedTest/java/org/oppia/android/app/topic/lessons/TopicLessonsFragmentTest.kt:321]: I added these again cause orientation does'nt change in Espresso
even if using `@config`

**Comment by BenHenning** [comment]
Thanks @FareesHussain. Can you actually change the test such that:
- We have a test for doing X (in portrait by default)
- We have another test for doing X after a configuration change
- We have another test for doing X (in landscape by default)

The third test is technically a new, duplicate test of the first one (but starting in landscape). This is preferred over changing existing tests to only be landsdcape + configuration change, and should improve our behavioral coverage a bit.

It looks like two tests were changed before--in that case I'd expect those two to be reverted, and duplicated to be the third class of test above (starting in landscape but performing a configuration change).

**Review by BenHenning — APPROVED** [review]
Thanks @FareesHussain! LGTM.

**Pull request merged**